Issue 47. Model classes for authorization DB tables

sbryzak opened the pull request:
This pull request contains changes for [ISSUE-47](https://github.com/fabric8-services/fabric8-auth/issues/47):

* New model classes for authorization services - resources, roles
* Migration DDL script for creation of new tables
* Unit tests for model classes (ongoing)

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-1`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-2`

alexeykazakov commented on authorization/resource/resource.go:
```diff
@@ -0,0 +1,223 @@
+		}, "unable to update the resource")
+		return errs.WithStack(err)
+	}
+	err = m.db.Model(obj).Updates(model).Error
```
I think using m.db.Save(&user) is safer than using Updates() in our Save() methods.
Save() updates all fields but Updates() updates only changed and not blank(!). So if resource.Description is an empty string then it wont be updated!
See more details here: http://jinzhu.me/gorm/crud.html#update

Overall our space resource repo https://github.com/fabric8-services/fabric8-auth/blob/master/space/space_resource.go is probably a better example of CRUD methods.

alexeykazakov commented on account/identity_relationship.go:
```diff
@@ -0,0 +1,189 @@
+	Save(ctx context.Context, u *IdentityRelationship) error
+	List(ctx context.Context) ([]IdentityRelationship, error)
+	Delete(ctx context.Context, parentIdentityID uuid.UUID, childIdentityID uuid.UUID) error
+	Query(funcs ...func(*gorm.DB) *gorm.DB) ([]IdentityRelationship, error)
```
I see there are Query() method in all repositories in this PR. Do you really need them everywhere? Are they all used?

sbryzak replied:
I'm actually thinking we should have more sensible "query" methods that reflect our likely use cases.  For example, it makes sense to be able to list the ResourceTypeScope records for a particular ResourceType, so we could add a ResourceTypeScope.List(ResourceType) method for that.  Same goes with listing Role objects for a particular ResourceType.  What do you think?

alexeykazakov replied:
Sure. We can add whatever methods we need. My point was mostly about not adding anything that we are not sure if we need now.

alexeykazakov commented on authorization/resource/resource.go:
```diff
@@ -0,0 +1,223 @@
+}
+
+// First returns the first Resource element that matches the given criteria
+func (m *GormResourceRepository) First(funcs ...func(*gorm.DB) *gorm.DB) (*Resource, error) {
```
The same question as "Query". Is this First() method is going to be used? I would suggest to add only CRUD methods that we will need for sure right now. Create(), Load(), Save(), Delete() and others only if needed. If not sure right now then let's not add it. We can add it later if needed.

alexeykazakov commented on authorization/resource/resource.go:
```diff
@@ -0,0 +1,223 @@
+	Save(ctx context.Context, resource *Resource) error
+	Delete(ctx context.Context, id string) error
+	Query(funcs ...func(*gorm.DB) *gorm.DB) ([]Resource, error)
+	List(ctx context.Context) ([]Resource, error)
```
There is no List() implementation which makes sense. So, remove List() from interface too.

alexeykazakov commented on design/resource.go:
```diff
@@ -0,0 +1,104 @@
+		a.Response(d.TemporaryRedirect)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.BadRequest, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
```
We could return Forbidden too I guess?

sbryzak replied:
I've added Forbidden.

alexeykazakov commented on design/resource.go:
```diff
@@ -0,0 +1,104 @@
+		a.Routing(
+			a.GET(""),
+		)
+		a.Description("List all resources")
```
We are not going to allow to list all the resources, are we? :)

sbryzak replied:
This method is required to be UMA compliant, however I'm thinking that we should just have it return an empty list.  What do you think?  Actually, we should probably support listing resources by owner.

alexeykazakov replied:
Let's not add this right now. We can add an empty list and other lists later if/when we need.

sbryzak replied:
No problem, I've removed the endpoint definition.

alexeykazakov commented on design/resource.go:
```diff
@@ -0,0 +1,104 @@
+	a.Description("A Protected Resource")
+	a.Attributes(func() {
+		a.Attribute("resource_scopes", a.ArrayOf(d.String), "The valid scopes for this resource")
+		a.Attribute("icon_uri", d.String, "The URI containing the icon for the resource")
```
Do we really need any icons here? I know that UMA specify them. But shouldn't we just ignore them?

sbryzak replied:
My opinion is that we allow the parameter in order to be compliant with UMA (while compliance might not be important right now it might be in the future) however we just don't do anything with it if one is passed in a request. So yes, we just ignore it.

alexeykazakov replied:
Sure. I'm not saying we should forbid this param. But we don't have to declare this param to allow it.  It will be allowed anyway. But by declaring it we generate additional code (this param will be included in goa model, context, tests, etc).

sbryzak replied:
So a JSON document passed in the request body does not have to conform with the MediaType definition declared for that method?  Won't there be some kind of validation error if the document contains properties that the MediaType does not declare?

alexeykazakov replied:
Oh, didn't notice it's a media type. It's not requests parameters. But anyway additional (not declared) attributes are just ignored. So, it's safe not to declare them. Payloads with undeclared attributes are still valid.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-3`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-4`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-5`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-6`

alexeykazakov commented on migration/sql-files/004-authorization.sql:
```diff
@@ -0,0 +1,70 @@
+
+CREATE TABLE role_scope (
+    resource_type_scope_resource_type_scope_id uuid references resource_type_scope(resource_type_scope_id),
+    role_role_id uuid references role(role_id),
```
It would be "role_id" if we use "id" as primary key in "role" table instead of "role_id", correct?

sbryzak replied:
Correct, but only because of gorm's naming convention defaulting to table_name_column_name without there being any possible way of overriding it.  This means that gorm would be totally broken for legacy databases in which it is not possible to change column names.

sbryzak replied:
Oh, and other people are experiencing this same issue:

https://stackoverflow.com/questions/33437453/using-different-join-table-columns-than-defaults-with-gorm-many2many-join

alexeykazakov replied:
So, another reason to use short column names for primary keys - "id"? :) I understand that's a gorm limitation but columns like resource_type_scope_resource_type_scope_id doesn't look good at all.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-7`

sbryzak commented:
Ok, I think I've addressed everything now.  After a lot of messing around and refactoring to compensate for bugs in gorm, I'm pretty happy with what's in this patch.  I've moved the identity_relationship stuff out of this issue and will implement it in a future issue as it is quite a comprehensive feature request.   I've also been able to fix the crazy column names for the ROLE_SCOPE table to something much more sensible.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-14`

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-15`

sbryzak commented:
The build failed with this error, which is caused by something outside the scope of this patch:

=== RUN   TestShowUserOKUsingExpiredIfModifedSinceHeader
time="2017-08-31T12:43:40Z" level=info msg="Identity created!" identity_id=b4a38d8b-0b4a-4fac-897c-d45ad46d28a6 pkg=account req_id= 
time="2017-08-31T12:43:40Z" level=info msg="Deleting entities from 'identities' table with key b4a38d8b-0b4a-4fac-897c-d45ad46d28a6" hook_name="mighti:record-27fd9b08-627b-437d-a94b-d3122c6a7967" key=b4a38d8b-0b4a-4fac-897c-d45ad46d28a6 pkg="gormsupport/cleaner.DeleteCreatedEntities" table=identities 
time="2017-08-31T12:43:40Z" level=info msg="Deleting entities from 'users' table with key 86c3576f-7f0c-41d5-b7d1-e020bb380fe1" hook_name="mighti:record-27fd9b08-627b-437d-a94b-d3122c6a7967" key=86c3576f-7f0c-41d5-b7d1-e020bb380fe1 pkg="gormsupport/cleaner.DeleteCreatedEntities" table=users 
[info] removing callback `mighti:record-27fd9b08-627b-437d-a94b-d3122c6a7967` from /tmp/go/src/github.com/fabric8-services/fabric8-auth/gormsupport/cleaner/db_clean.go:78
--- FAIL: TestShowUserOKUsingExpiredIfModifedSinceHeader (0.01s)
	assertions.go:241: 
                        
	Error Trace:	users_blackbox_test.go:906
		
			users_blackbox_test.go:595
		
	Error:		Not equal: "tQz6146/hefpro+o3m6fOg==" (expected)
			        != "sd3AKmSL8ahq5AHI2Ca7jA==" (actual)

I'm wondering if it is an intermittent problem, because I think I have seen this happen once before locally.  Is there any way to restart the build again without pushing another commit?

sbose78 commented:
[test]

alexeykazakov commented:
[test]

alexeykazakov commented:
That test error is inherited problem from WIT. Something is broken (in the test???) with that "if-modified" cache thing. It's not related to this PR.

alexeykazakov commented on authorization/resource/resource_type_scope.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// ResourceTypeScopeFilterByID is a gorm filter for Resource Type Scope ID.
+func ResourceTypeScopeFilterByID(resourceTypeScopeID uuid.UUID) func(db *gorm.DB) *gorm.DB {
```
Do we need it?

alexeykazakov commented on authorization/resource/resource_type_scope_blackbox_test.go:
```diff
@@ -0,0 +1,154 @@
+
+func (s *resourceTypeScopeBlackBoxTest) TestOKToDelete() {
+	t := s.T()
+	res.Require(t, res.Database)
```
You don't have to require Database explicitly in gormtestsupport.DBTestSuite test suites. It's already required.

alexeykazakov commented on authorization/resource/resource_type_scope_blackbox_test.go:
```diff
@@ -0,0 +1,154 @@
+
+	resourceType := &resource.ResourceType{
+		ResourceTypeID: uuid.NewV4(),
+		Name:           "Area" + uuid.NewV4().String(),
```
Let's add the test name to all Names of all records created by tests. If the test doesn't clean the created records then it will be easy to identify what test is responsible for that. It's good to use in all DB tests.

alexeykazakov commented on authorization/role/identity_role_blackbox_test.go:
```diff
@@ -0,0 +1,194 @@
+
+func (s *identityRoleBlackBoxTest) TestOKToDelete() {
+	t := s.T()
+	res.Require(t, res.Database)
```
Not needed.

alexeykazakov commented on authorization/role/role.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+// RoleFilterByID is a gorm filter for Role ID.
+func RoleFilterByID(roleID uuid.UUID) func(db *gorm.DB) *gorm.DB {
```
It's not used, right?

alexeykazakov commented on authorization/role/role_blackbox_test.go:
```diff
@@ -0,0 +1,182 @@
+
+func (s *roleBlackBoxTest) TestOKToDelete() {
+	t := s.T()
+	res.Require(t, res.Database)
```
You can remove it.

fabric8cd commented:
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-68-16`

sbryzak commented:
[test]

sbryzak commented:
[test]

sbryzak commented:
I keep getting test failures in the CI build, the latest one is this:

=== RUN   TestCreateAndDeletePermissionOK
time="2017-09-01T12:00:10Z" level=error msg="unable to obtain token" file="/tmp/go/src/github.com/fabric8-services/fabric8-auth/controller/token.go" func="github.com/fabric8-services/fabric8-auth/controller.GenerateUserToken" line=198 pid=10246 pkg=controller req_id= response_body="{\"error\":\"invalid_grant\",\"error_description\":\"Invalid user credentials\"}" response_status="401 Unauthorized" 
--- FAIL: TestCreateAndDeletePermissionOK (0.40s)
	assertions.go:241: 
                        
	Error Trace:	authz_blackbox_test.go:492
		
			authz_blackbox_test.go:437
		
			authz_blackbox_test.go:137
		
	Error:		Expected nil, but got: errors.InternalError{Err:unable to obtain toke. Response status: 401 Unauthorized. Responce body: }

alexeykazakov commented:
[test]

alexeykazakov commented:
[test]

alexeykazakov commented:
It was a mess in our test realm in KC. I cleaned it up now it should be better. We need to fix https://github.com/fabric8-services/fabric8-auth/issues/13 to prevent such failures.
